ouroboros-consensus/test/consensus-test/Test/Consensus/MiniProtocol/ChainSync/Client.hs

```diff
@@ -312,99 +412,164 @@ runChainSync securityParam (ClientUpdates clientUpdates)
-           varCandidates
-           serverId
-           maxBound $ \varCandidate -> do
-             atomically $ modifyTVar varFinalCandidates $
-               Map.insert serverId varCandidate
-             result <-
-               runPipelinedPeer protocolTracer codecChainSyncId clientChannel $
-                 chainSyncClientPeerPipelined $ client varCandidate
-             atomically $ writeTVar varClientResult (Just (Right result))
-             return ()
-        `catchAlsoLinked` \ex -> do
-          atomically $ writeTVar varClientResult (Just (Left ex))
-          -- Rethrow, but it will be ignored anyway.
-          throwIO ex
-      void $ forkLinkedThread registry "ChainSyncServer" $
-        runPeer nullTracer codecChainSyncId serverChannel
-                (chainSyncServerPeer server)
-
-    LogicalClock.waitUntilDone clock
-    -- Wait a random amount of time after the final tick for the chain sync
-    -- to finish
-    threadDelay 2000
+    --
+    -- Happens /immediately after/ the chain and clock effects schedule for
+    -- 'startSyncingAt'.
+    let initiateChainSync = do
+            (clientChannel, serverChannel) <- createConnectedChannels
+            -- Don't link the thread (which will cause the exception to be
+            -- rethrown in the main thread), just catch the exception and store
+            -- it, because we want a "regular ending".
+            void $ forkThread registry "ChainSyncClient" $
+              bracketChainSyncClient
+                 chainSyncTracer
+                 chainDbView
+                 varCandidates
+                 serverId
+                 maxBound $ \varCandidate -> do
+                   atomically $ modifyTVar varFinalCandidates $
+                     Map.insert serverId varCandidate
+                   result <-
+                     runPipelinedPeer protocolTracer codecChainSyncId clientChannel $
+                       chainSyncClientPeerPipelined $ client varCandidate
+                   atomically $ writeTVar varClientResult (Just (ClientFinished result))
+                   return ()
+              `catchAlsoLinked` \ex -> do
+                atomically $ writeTVar varClientResult (Just (ClientThrew ex))
+                -- Rethrow, but it will be ignored anyway.
+                throwIO ex
+            void $ forkLinkedThread registry "ChainSyncServer" $
+              runPeer nullTracer codecChainSyncId serverChannel
+                      (chainSyncServerPeer server)
+
+    -- If the candidate's tip's slot's onset is ahead of the local wall-clock
+    -- (which is skewed by 'clientSlowBy'), then the ChainSync client
+    -- mishandled a block from the future.
+    let checkTipTime :: m ()
+        checkTipTime = do
+            (now, candidates) <- do
+              now        <- systemTimeCurrent systemTime
+              candidates <-
+                atomically $ readTVar varCandidates >>= traverse readTVar
+              pure (now, candidates)
```

Comment:
```suggestion
            now        <- systemTimeCurrent clientSystemTime
            candidates <-
              atomically $ readTVar varCandidates >>= traverse readTVar
```

Reply:
Thanks: it had been a common `atomically` before a refactor 👍